### ISSUE TITLE
Replace uses of tempname in CI by mktemp

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#43597]
As of #38879 our temp path generation depends on the global RNG. That's probably not a bad idea in general, except that various parts of the test suite seed the global RNG, so this will cause reliable collisions in the test suite, so we'll get in trouble with tests stomping on each other. We should audit all uses of `tempname` in the test suite and replace them with mktemp unless they test the tempname function specifically.

**Comment by staticfloat** [comment]
I thought that `Libc.rand()` explicitly doesn't get seeded?

**Comment by Keno** [comment]
You're right. It does use Libc.rand. And now that I'm looking back at the two reports I linked in #38879, I also see different filenames, so I might be going crazy or I linked the wrong reports.... Sigh. Regardless, we should probably still get rid of tempname.

**Comment by Keno** [comment]
Related: The `TEMP_CLEANUP` stuff is pretty dangerous because it some other process could have come in and created a new file with the same name at that location, and we just indiscriminately delete it.

**Comment by vtjnash** [comment]
> pretty dangerous because it some other process could have come in and created a new file with the same name at that location, and we just indiscriminately delete it

That is exactly equivalent to the original problems of `tempnam` . See the first note at https://linux.die.net/man/3/tempnam.

**Comment by DilumAluthge** [comment]
> Related: The `TEMP_CLEANUP` stuff is pretty dangerous because it some other process could have come in and created a new file with the same name at that location, and we just indiscriminately delete it.

@StefanKarpinski How do you think we should approach this?

One option would be to tell people (in the documentation) that if they do `mktemp(; cleanup = true)` or `mktempdir(; cleanup = true)`, they should never manually delete the temp file/directory, because Julia will do it automatically when the Julia process ends.

A second option, which might be better, would be to add two new functions `rmtemp` and `rmtempdir`, which users can use if they want to manually delete a temp file/directory at some point before the Julia process ends. When the `rmtemp`/`rmtempdir` function is called, after it deletes the temp file/directory, it also removes the temp filename/directory name from the list of names that Julia is going to cleanup when the Julia process ends. So then, when Julia exits, it doesn't try to delete those names, because we removed those names from the list already.

A third option would be to basically do the same thing as the second option, but instead of making separate `rmtemp`/`rmtempdir` functions, we just hook it into the existing `rm` functionality.

**Comment by StefanKarpinski** [comment]
I think that tempname generation should not depend on the same RNG that can be seeded. It could use `RandomDevice` or some independent RNG that we don't expose for people to seed.

**Comment by StefanKarpinski** [comment]
Should we use that for `RandomDevice` everywhere as well?

**Comment by vtjnash** [comment]
Probably. It gives access to some better kernel APIs on most systems to read system randomness (e.g. the getrandom syscall).

**Comment by vtjnash** [comment]
Partly addressed by https://github.com/JuliaLang/julia/pull/44432 now, though the TEMP_CLEANUP tests still fail pretty often, and cause the Distributed workers to run out of time at exit now
